### PR TITLE
accesslog: Append a newline to returned log_lines in access_log

### LINKED ISSUE #5193: question: debugging access log format problems
*Title*: question: debugging access log format problems

*Description*:
The access log formatter does not appear to work without a null terminator \n . #87 was closed with an [update to the docs](https://github.com/envoyproxy/envoy/pull/93/files) noting the need for a newline separator. It makes sense to not enforce the newline when inserting a custom format, but a colleague and I found it difficult to determine that this may have been the problem since no leveled logging printed that there may be a problem. 

I also tried validating with the [config-checker tool](https://github.com/envoyproxy/envoy/blob/master/test/tools/config_load_check/config_load_check.cc) , but it did not complain about the missing \n. 

The JSON logger now automatically adds newlines with #5024 but the regular logger still can fail silently due to a bad format. Are there any potential logs or warnings that should catch these types of format problems, and if not would a PR to add some trace logs around this validation be accepted?

*Repro steps*:
1. Use curl get against an Envoy proxied web address with the custom access logger listed below and log-level set to trace.
2. Note that even with trace logging enabled, no warnings are printed and the access log is not displayed at all.

*Config*:

```yaml
format: "[%START_TIME%] %PROTOCOL% %BYTES_RECEIVED% %BYTES_SENT% %RESPONSE_FLAGS% %UPSTREAM_HOST% %UPSTREAM_CLUSTER% %REQUESTED_SERVER_NAME%"
```

### PR DESCRIPTION
**Pull request by cmluciano**
Signed-off-by: Christopher M. Luciano <cmluciano@us.ibm.com>

*Description*: Append a newline to returned log_line in access_log
*Risk Level*: Medium - Might be worked around by external consumers already that add a newline themselves
*Testing*: Adjusted tests to account for newline 
*Docs Changes*: TODO
*Release Notes*: TODO
Fixes #5193 

Notes for reviewer: 
- This change has the unfortunate side-effect of [breaking the verifyJsonOutput test](https://github.com/envoyproxy/envoy/blob/master/test/common/access_log/access_log_formatter_test.cc#L386-L399) since it expects only one ending terminator and does not expect the first pair to contain a newline. 
- Is it sufficient to drop some of these expects and update the strings in the JSON tests or have I gone about adding this the wrong way?

### CODE REVIEW

File: source/common/access_log/access_log_formatter.cc [hunk 1]
```diff
@@ -68,7 +68,7 @@ std::string FormatterImpl::format(const Http::HeaderMap& request_headers,
     log_line += provider->format(request_headers, response_headers, response_trailers, stream_info);
   }
 
-  return log_line;
+  return absl::StrCat(log_line, "\n");
```
Comment (ggreenway) [inline source/common/access_log/access_log_formatter.cc]: I think you need to condition this on whether it already ends in a newline. Also, it'll be more efficient to do that when constructing the formatter. See AccessLogFormatParser::parse(). I'd just add a PlainStringFormatter to the end of the vector that contains a `\n`, but only do that if it does not already end in `\n`

Comment (cmluciano) [reply source/common/access_log/access_log_formatter.cc]: Excellent! I was hoping that I was just missing something obvious here.

**Comment by cmluciano** [comment]
@ggreenway Is it still expected that I should have to update the json stubs to expect a newline for the [first.pair](https://github.com/envoyproxy/envoy/blob/master/test/common/access_log/access_log_formatter_test.cc#L396-L398)?

**Comment by ggreenway** [comment]
I haven't looked closely at how json logs are output. I thought they were using the protobuf-->json formatter; I'm unsure if they share code with the "normal" text access logger. So I don't know the answer to your question.

**Comment by cmluciano** [comment]
not stale, just working on other PRs first

**Pull request closed without merging**